python/lsst/ts/m1m3/utils/thermocouples.py

```diff
@@ -52,318 +55,1000 @@ async def get_scanner_data(
+
+            # Calculate r gradients
+            ArTAr = Ar.T @ Ar
+            ArTy = Ar.T @ temperature
+
+            beta = np.linalg.lstsq(ArTAr, ArTy, rcond=None)[0]
+
+            y_hat = Ar @ np.linalg.lstsq(Ar, temperature, rcond=None)[0]
+            dof = max(len(y) - Ar.shape[1], 1)
+            sigma2 = float(np.sum((y - y_hat) ** 2) / dof)
+            cov = sigma2 * np.linalg.pinv(ATA)
+            errs = np.sqrt(np.diag(cov))
+
+            all_gr.append(beta[1])
+            all_gr_err.append(errs[1])
+
+        if use_3d_dataset:
+            return pd.DataFrame(
+                data={
+                    "intercept": intercepts,
+                    "intercept_err": intercepts_err,
+                    "x_gradient": all_gx,
+                    "x_gradient_err": all_gx_err,
+                    "y_gradient": all_gy,
+                    "y_gradient_err": all_gy_err,
+                    "z_gradient": all_gz,
+                    "z_gradient_err": all_gz_err,
+                    "radial_gradient": all_gr,
+                    "radial_gradient_err": all_gr_err,
+                },
+                index=temperatures.index,
+            )
+
+        else:
+            return pd.DataFrame(
+                data={
+                    "intercept": intercepts,
+                    "intercept_err": intercepts_err,
+                    "x_gradient": all_gx,
+                    "x_gradient_err": all_gx_err,
+                    "y_gradient": all_gy,
+                    "y_gradient_err": all_gy_err,
+                    "radial_gradient": all_gr,
+                    "radial_gradient_err": all_gr_err,
+                },
+                index=temperatures.index,
+            )
+
+    def compute_temp_stats_and_rate(
+        self,
+        data: pd.DataFrame,
+        window_minutes: float = 30,
+    ) -> pd.DataFrame:
+        """
+        Compute per-timestamp temperature stats and a windowed rate of change.
+
+        Parameters
+        ----------
+        data : pd.DataFrame
+            Index: pandas.DatetimeIndex (tz-aware or naive).
+            Columns: temperature sensors. Values are temperatures (any units).
+        window_minutes : float
+            Size of the look-back window (e.g., 30 for 30 minutes).
+
+        Returns
+        -------
+        pd.DataFrame
+            Columns:
+            - 'mean_temp'   : mean across sensors (row-wise)
+            - 'std_temp'    : std across sensors (row-wise)
+            - 'range_temp'  : (row-wise max - min)
+            - f'rate_per_{window_minutes}min': d(mean)/dt over the window,
+                in temperature-units per minute. Rows lacking enough history
+                are NaN. Length matches len(df) and index matches df.index.
+        """
+
+        if not isinstance(data.index, pd.DatetimeIndex):
+            raise TypeError("DataFrame index must be a pandas.DatetimeIndex.")
+
+        # Row-wise stats (preserve original index order, allow NaNs)
```

Comment:
These stats all look pretty choppy when plotting, expecially when using the rate_per_window...I think we might be suffering from some outlier noise. We could try using something like 90/10 ranges and MAD instead of min/max and std.